### PR TITLE
Allow PATCHes to succeed on readOnly and immutable fields when not modified

### PR DESCRIPTION
**Pull request by 11xor6**
While doing an integration with a large online IdP it came up that they would PATCH `readOnly` fields of resources with unchanged values. This isn't strictly against the spec (from https://tools.ietf.org/html/rfc7644#section-3.5.2):
> Each operation against an attribute MUST be compatible with the
attribute’s mutability and schema as defined in Sections 2.2 and 2.3
of [RFC7643].  For example, a client MUST NOT modify an attribute
that has mutability “readOnly” or “immutable”.  […] An operation that
 is not compatible with an attribute’s mutability or schema SHALL
return the appropriate HTTP response status code and a JSON detail
error response as defined in Section 3.12.

It's technically not doing anything wrong here because nothing is being modified. While I don't particularly agree with their implementation, I do need to work with it. Happy to discuss or consider other implementations.

### CODE REVIEW

**Comment by Captain-P-Goldfish** [comment]
The solution for immutable fields  is ok but I do not think the readOnly fields should be mutable by a client. There is a good reason for having a distinct meaning between readOnly and immutable. But with this solution readOnly will actually have the same meaning as immutable.
The resource definitions can be changed easily by modifying the schema definitions. Wouldn't it be a better solution for you to turn the affected fields to mutability immutable?

**Comment by 11xor6** [comment]
Actually the purpose here isn't to change the dynamics of what immutable and readOnly mean, but to ignore `PATCH`es to fields of those types if they don't actually change any values. Trying to change the value of a readOnly or a previous written immutable still fails.

**Comment by 11xor6** [comment]
For example if I have a `Group` with ID `1234` and I receive a `PATCH` that's modify ID to be `1234` (no change) this just ignores that. However If the `PATCH` is modify ID to be `5678` that will fail with a 409 (Conflict).

File: scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java [hunk 1]
```diff
@@ -89,24 +89,57 @@ public boolean addResourceValues(ObjectNode resource, JsonNode readJsonDocument,
...
     return changeWasMade.get();
   }
 
+  /**
+   * determines if the given json node is read-only or immutable and previously set, if so returns true
+   *
+   * @param jsonNode the node that should be verified
+   * @param schemaAttribute the schema definition of the node
+   */
+  private boolean isReadOnly(JsonNode jsonNode, SchemaAttribute schemaAttribute)
```
Comment (Captain-P-Goldfish) [inline scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java:119]: please rename method to: 

isReadOnlyAndSet

File: scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java [hunk 2]
```diff
@@ -89,24 +89,57 @@ public boolean addResourceValues(ObjectNode resource, JsonNode readJsonDocument,
       else
       {
         SchemaAttribute schemaAttribute = getSchemaAttribute((extensionUri == null ? "" : extensionUri + ":") + key);
-        verifyImmutableAndReadOnly(resource, schemaAttribute);
+        boolean resourceChanged;
+        boolean isReadOnly = isReadOnly(resource, schemaAttribute);
```
Comment (Captain-P-Goldfish) [inline scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java:93]: please rename fielt to:

isReadOnlyAndSet

File: scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java [hunk 3]
```diff
@@ -89,24 +89,57 @@ public boolean addResourceValues(ObjectNode resource, JsonNode readJsonDocument,
...
+  }
+
+  /**
+   * verifies that the given json node is neither readOnly nor immutable and already set
+   *
+   * @param schemaAttribute the schema definition of the node
+   * @param resourceChanged true if the resource was modified
+   * @param isReadOnly true if the field is read-only OR immutable and previously set
+   */
+  private void verifyReadOnlyNotModified(SchemaAttribute schemaAttribute, boolean resourceChanged, boolean isReadOnly)
```
Comment (Captain-P-Goldfish) [inline scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java:133]: please rename parameter "isReadOnly" to "isReadOnlyAndSet"

**Comment by Captain-P-Goldfish** [comment]
The misunderstanding was on my side. What you did is perfectly fine. :-)

**Comment by Captain-P-Goldfish** [comment]
by the way, so just that you know:
I just noticed that the delombok commit caused problems with debugging since source and byte code did not match. I moved it to the release-only profile. Like this it will be generated correctly for releases and does not affect the developing process.

**Comment by 11xor6** [comment]
I'll make those changes, give me two minutes. 

As for the delombok, totally makes sense; I did my best to make sure it worked, but I'm glad you caught that.

File: scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java (hunk 1 above)
Comment (11xor6) [reply scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java]: 👍

File: scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java (hunk 2 above)
Comment (11xor6) [reply scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java]: 👍

File: scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java (hunk 3 above)
Comment (11xor6) [reply scim-sdk-server/src/main/java/de/captaingoldfish/scim/sdk/server/patch/PatchResourceHandler.java]: 👍

**Pull request merged**